### ISSUE TITLE
REINFORCE call in the decoder base class needs a detach call on top of the reward.

### ISSUE THREAD

**Issue by prithv1** [batra-mlp-lab/visdial-rl#13]
https://github.com/batra-mlp-lab/visdial-rl/blob/1fb7e88a65013477b0eb8f14dec2fda74eb48d17/visdial/models/decoders/gen.py#L243

Following the paper, the above should be replaced by 
`loss += -1 * log_prob * (reward.detach() * (self.mask[:, t].float()))`

Not having a `.detach()` on the reward here provides another source of gradients to the feature regression module in addition to the feature loss, the only difference being these gradients are scaled by the log-probs, which does not seem to mean anything intuitively.
